Add I/O for links' TS generation

flomnes opened the pull request:
Add read for
- "modulation" timeseries => `input/links/<link 1>/prepro/<link 2>_{direct, indirect}.txt`
- "prepro" timeseries => `input/links/<link 1>/prepro/<link 2>_mod_{direct, indirect}.txt`
- link parameters => `tsgen_direct_XXX`, `tsgen_indirect_XXX`, with XXX = unitcount, nominalcapacity, etc.
- number of TS to generate => `generaldata.ini/General/nbtimeserieslinks`

Add `bool` argument to optionally load time-series. This bool is `false` (default value) in `antares-solver`, and `true` in `antares-ts-generator`. Parameters are loaded in any case.

De-couple `PreproThermal` and `ThermalCluster` (remove the dependency of the 1st to the 2nd).

flomnes committed:
Add I/O for links' TS generation

flomnes committed:
Fix memory leak

guilpier-code committed:
STS groups : make group categories dynamic rather than fixed  (#2007)

We make groups of **S**hort **T**erm **S**torage dynamic. 
This means that accepted groups are no longer fixed in the code.
Instead, any group can be defined by user.
It has consequences on the **Thematic Trimming**, which now accepts only
one item (namely **STS BY GROUP**) regarding STS groups output
variables.
Turning it to on prints all variables related to groups. 

Note : 
- We added a default group name (in case user doesn't specify a group
name for a cluster) : "**sts-def-group**". Feel free to amend it.
- A new test (a study) was created. Please find it
[here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/files/14802896/ST-dynamic-groups.zip).
Look at its **user's notes** to know its intentions.
- upon a @sylvlecl suggestion, doc was updated accordingly

## Required
- [x] Merge
https://github.com/AntaresSimulatorTeam/Antares_Simulator_Tests_NR/pull/18/files,
generate new reference results

---------

Co-authored-by: Florian Omnès <26088210+flomnes@users.noreply.github.com>

flomnes committed:
Extract functions for STS groups (#2020)

flomnes committed:
Merge remote-tracking branch 'github/feature/thermal-interface' into feature/links-gen-io

flomnes committed:
Lower case for properties names

flomnes committed:
Fix success

flomnes committed:
Use SimTest@v9.1.0-rc3 (#2021)

- Fix thematic trimming variable names for short-term-storage by
@flomnes in
https://github.com/AntaresSimulatorTeam/Antares_Simulator_Tests_NR/pull/18
- Add a study with dynamic STS groups (v9.1.0) by @flomnes in
https://github.com/AntaresSimulatorTeam/Antares_Simulator_Tests_NR/pull/19

flomnes committed:
Possible fix for segfault (to be checked)

flomnes committed:
Use std::unique_ptr<>

flomnes committed:
Merge remote-tracking branch 'github/feature/thermal-interface' into feature/links-gen-io

flomnes committed:
const ref

flomnes committed:
Add I/O for Parameters::nbLinkTStoGenerate

flomnes committed:
Add format documentation

guilpier-code commented on docs/reference-guide/13-file-format.md:
```diff
@@ -1,6 +1,25 @@
+- volatility.planned (double in [0,1])
+- volatility.forced (same)
+
+- "modulation" timeseries => input/links/<link 1>/prepro/<link 2>_{direct, indirect}.txt, 365x6 values, respectively "forced outage duration", "planned outage duration", "forced outage rate", "planned outage rate", "minimum of groups in maintenance", "maximum of groups in maintenance".
```
There may be a confusion here, I guess : 
- **prepro** data file is supposed to contain the 6 columns
- **modulation** data file is supposed to contain the sole column (and have **\_mod\_** in its name)

payetvin replied:
done

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -471,7 +540,15 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
         link.displayComments = true;
 
-        ret = link.loadTimeSeries(study, folder) && ret;
+        if (loadTSGen)
```
There are thermal consideration here, in the middle of a link loading function.
As said above, we should do some renaming.

payetvin replied:
I don't see thermal consideration here, loadTSGen is used to know if we're loading the links via the solver or the ts generator tool

guilpier-code replied:
The type **PreproThermal** is a thermal consideration, is it not ?
About **loadTSGen**, see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2019#discussion_r1553263253)

flomnes committed:
Use id & unitCount instead of weak_ptr<ThermalCluster>

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -132,6 +133,35 @@ bool AreaLink::linkLoadTimeSeries_for_version_820_and_later(const AnyString& fol
     filename.clear() << capacitiesFolder << SEP << with->id << "_indirect.txt";
     success = indirectCapacities.loadFromFile(filename, false) && success;
 
+    if (loadTSGen)
```
Please see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2019#discussion_r1553263253) first.

`if (loadTSGen)` (and the parameter **loadTSGen**)  : 

To avoid to make this function more complex with this new **if** statement, we should create a new method with the new loading ability, and remove **loadTSGen** as argument. 
We could eventually have (one level higher) : 
```C++
bool AreaLink::loadTimeSeries(const StudyVersion version, const AnyString& folder, bool loadTSGen)
{
    if (version < StudyVersion(8, 2))
       return loadCapacities_below_v810(folder) && 
              loadParameters_below_v810(folder);
    
    // Loading data of format over v820
    ret = loadCapacities_over_v820(folder); // No more loadTSGen argument here
    ret = loadParameters_over_v820(folder) && ret;
    if (loadTSGen)
       ret = loadTSgenerator(folder) && ret;
}
```
Please note :  
- As we're in class **AreaLink**, we don't need to repeat **link** in methods' names
- loadTSGen disappears from loading functions calls, but is used one level higher.

flomnes replied:
Done !

flomnes committed:
Merge remote-tracking branch 'github/feature/thermal-interface' into feature/links-gen-io

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -111,7 +112,7 @@ bool AreaLink::linkLoadTimeSeries_for_version_below_810(const AnyString& folder)
 }
 
-bool AreaLink::linkLoadTimeSeries_for_version_820_and_later(const AnyString& folder)
+bool AreaLink::linkLoadTimeSeries_for_version_820_and_later(const AnyString& folder, bool loadTSGen)
```
**bool loadTSGen** : 
As you may already know,  if legacy code wasn't what it is, we would **not** need this new parameter at all.
We would only call the code that reads TS generations parameters for links from the **TS generator**, and we wouldn't call it from the **solver**, or the **cleaner**.
Instead, we have many occurrences of this parameter here in the reading functions, and some others at the higher levels. 

At some point, we should :
- separate data loading from : 
   - data checking and correcting 
   - business rules objects creation.
- make objects dedicated to loading data  (for example for a link) a collaboration of smaller, more individual, loading data object (for properties, for TS about capacities, for TS about modulation, ...), each one individually  usable in any context.

Doing this, we could load data on demand, for instance from the **TS generator**.

flomnes replied:
Good catch ! Done and it indeed simplifies the `loadTimeSeries` function a lot !

guilpier-code replied:
> Done and it indeed simplifies the `loadTimeSeries` function a lot !

I'm not sure it's done  : we still have occurrences of the parameters **loadTSGen**. 
But doing what I recommend would require some amount of work.

flomnes replied:
Yes, I don't want to go too deep here

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -471,7 +548,20 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
+              id_indirect, link.tsGenerationIndirect.unitCount);
+        }
+
+        ret = link.loadTimeSeries(study, folder, loadTSGen) && ret;
```
About `ret = link.loadTimeSeries(study, folder, loadTSGen) && ret;` : 
**study** is passed only for study version. We should pass the study version instead of the whole study.

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -406,7 +472,14 @@ bool AreaLinksInternalLoadFromProperty(AreaLink& link, const String& key, const
-
+    if (loadTSGen)
+    {
+        const std::string key_(key); // Conversion from Yuni::String
```
Converting the current function argument **const Yuni::String& key** into **std::string** would be easy.
We wouldn't need the **key_** local variable.
We could have a call like : 
```C++
if (!AreaLinksInternalLoadFromProperty(link, key.to<std::string>(), value, loadTSGen))
    ...
```
And why not convert the value parameter as well ?

flomnes replied:
You're right, my way is to avoid creating new instances of `Yuni::String`. However, it's not the point of this PR to migrate **existing** `Yuni::String`s into `std::string`s

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -406,7 +472,14 @@ bool AreaLinksInternalLoadFromProperty(AreaLink& link, const String& key, const
+    if (loadTSGen)
+    {
+        const std::string key_(key); // Conversion from Yuni::String
+        if (key_.starts_with("tsgen_direct"))
```
Same idea as above (splitting the catching properties function) : 

```C++
// Inifile
const IniFile::Property* p = s->firstProperty;
for (; p; p = p->next)
{
    key = p->key;
    key.toLower();
    value = p->value;
    bool found = catchLinkProperties(link, key, value));
    if (loadTSGen)
       found = catchLinkTSgenerationProperties(link, key, value) || found;
    if (!found) 
       logs.warning() << '`' << p->key << "`: Unknown property";
}
```
Function names were clarified and simplified.

flomnes replied:
```C++
// Inifile
const IniFile::Property* p = s->firstProperty;
for (; p; p = p->next)
{
    key = p->key;
    key.toLower();
    value = p->value;
    bool found = catchLinkProperties(link, key, value));
    if (!found && loadTSGen) // DIFF HERE
       found = catchLinkTSgenerationProperties(link, key, value);
    if (!found) 
       logs.warning() << '`' << p->key << "`: Unknown property";
}
```

flomnes replied:
done

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "antares/utils/utils.h"
 #include "antares/study/area/links.h"
```
About :
```C++
struct LinkTsGeneration
```
- Here we mix two different things : TS generation and business rules (links). This **struct** belongs more to TS generation, and should not be defined in this file.
- By the way, this **struct** could be used for a thermal cluster as well. So there is a kind of code duplication here.
- On the **Data::ThermalLaw** fields : should be renamed differently, as it is about statistic laws, and not specifically thermal laws.
- On the field `std::unique_ptr<Data::PreproThermal> prepro = nullptr;` : 
   - We don't need to initialize it to **nullptr**, it's done by default
   - about type **PreproThermal** : same as above. As it can be used for link capacities, it looses it's thermal-only characteristic, and should be renamed.

flomnes replied:
Done, except for PreproThermal (TODO)

flomnes replied:
Renamed `PreproThermal` -> `PreproAvailability`.

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -471,7 +548,20 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
 
-        ret = link.loadTimeSeries(study, folder) && ret;
+        if (loadTSGen)
+        {
```
About block : 
```C++
if (loadTSGen)
{
    ...
}
```
Too many implementation details we don't need to know about at this level.
We should call a function instead : 
```C++
if (loadTSGen)
    initializeTSgenerationPrepro(link);
```

guilpier-code replied:
Furthermore, here we initialize the **prepro**, and in a further step (right after) we load its data matrix.
As we're handling the **prepro** here, why not do it in one step ?
So for me this initialize step should be moved to the **link.loadTimeSeries(...)** step (see remark above).
This would remove an **if (loadTSGen)** block.

flomnes replied:
Yes, done. I had to check dependencies carefully

guilpier-code commented on src/libs/antares/study/cleaner/cleaner-v20.cpp:
```diff
@@ -359,7 +359,7 @@ bool listOfFilesAnDirectoriesToKeep(StudyCleaningInfos* infos)
                 // load all links
                 buffer.clear() << infos->folder << "/input/links/" << area->id;
-                if (not AreaLinksLoadFromFolder(*study, arealist, area, buffer))
+                if (not AreaLinksLoadFromFolder(*study, arealist, area, buffer, false))
```
New **bool** argument should not be needed. See [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2019#discussion_r1553263253)

flomnes replied:
I removed this argument from most functions, but it would be a bit more work to do it here

guilpier-code replied:
You're right, it would require more work.
But we should keep in mind that we don't need this **loadTSGen** fundamentally.
It's a workaround because the code has low quality.

flomnes replied:
Yes, the loading functions are not de-coupled from the validation functions.

guilpier-code commented on src/solver/ts-generator/thermal.cpp:
```diff
@@ -60,7 +60,7 @@ ThermalInterface::ThermalInterface(Data::AreaLink::LinkTsGeneration& source,
  forcedLaw(source.forcedLaw),
  plannedLaw(source.plannedLaw),
- prepro(source.prepro),
+ prepro(source.prepro.get()),
```
Maybe already said in the review of [PR 1986](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1986), but **thermal.cpp** is a bad name for this file.
It mostly contains code about TS generation (and a certain kind of TS generation).
So we should rename it for example **availabilityTSgeneration.cpp** (maybe not the right name, by the main idea is here).
The code unrelated to TS generation (for example code specific to links or thermal clusters) should be moved elsewhere

guilpier-code replied:
Not taken care of yet. 
Is there a problem to perform this change ?

flomnes replied:
[done](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2019/commits/b647031d0cf14beedfc0ac21afe6fdd5ef302c57)

payetvin committed:
comments 1

payetvin commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -170,12 +200,12 @@ void AreaLink::overrideTransmissionCapacityAccordingToGlobalParameter(
+bool AreaLink::loadTimeSeries(const StudyVersion &version, const AnyString& folder, bool loadTSGen)
 {
-    if (study.header.version < StudyVersion(8, 2))
+    if (version < StudyVersion(8, 2))
```
Add a case for version >= 9.1 ?
Currently handled by loadTSGen

flomnes committed:
PR remarks

flomnes committed:
Modify implementation (PR Remarks)

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/links-gen-io

flomnes committed:
Rename PreproThermal -> PreproAvailability

find src -regextype egrep -regex ".*/*\.(cpp|h|hxx)$"  -exec sed -i "s/PreproThermal/PreproAvailability/g" {} \;

flomnes committed:
Fix bug, simplify

flomnes committed:
key, link, value

flomnes committed:
Fix success

flomnes committed:
Optional loading for TS

flomnes committed:
Rename ThermalInterface -> AvailabilityTSGeneratorData

flomnes committed:
Fix log

flomnes committed:
Move thermal.cpp -> availability.cpp

payetvin committed:
code smells 1

guilpier-code commented on src/solver/ts-generator/include/antares/solver/ts-generator/prepro.h:
```diff
@@ -24,14 +24,15 @@
 ** \brief Thermal
 */
-class PreproThermal
+class PreproAvailability
```
**PreproAvailability** : watch for max value, which is **thermalPreproMax**

flomnes replied:
[done](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2019/commits/62a87063ff4648c0b45ba6c3830a4ac0509a0715)

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -406,20 +469,64 @@ bool AreaLinksInternalLoadFromProperty(AreaLink& link, const String& key, const
+    return false;
+}
+
+bool handleTSGenKey(Data::AreaLink& link, const std::string& key, const String& value)
```
Now arg **key** is an **std::string &**.
Do we need to define **key_** ?
If not, we should remove `const std::string key_(key);`

flomnes replied:
[done](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2019/commits/07278105c09907bf73a2d219f38cfb034f09e708)

guilpier-code commented on src/solver/ts-generator/include/antares/solver/ts-generator/prepro.h:
```diff
@@ -104,9 +110,27 @@ class PreproThermal
+    unsigned int unitCount;
+}; // class PreproAvailability
+
+struct LinkTsGeneration
```
I still believe that there is a kind of code duplication between this type **LinkTsGeneration** and its thermal cluster counterpart. 
But it also would be not immediate to make it disappear. So I understand if we give up for now.

flomnes replied:
Right

guilpier-code commented on src/solver/ts-generator/include/antares/solver/ts-generator/prepro.h:
```diff
@@ -104,9 +110,27 @@ class PreproThermal
+    unsigned int unitCount;
+}; // class PreproAvailability
+
+struct LinkTsGeneration
```
I still believe that there is some kind of code duplication between this type **LinkTsGeneration** and its thermal cluster counterpart. 
But it also would be not immediate to make it disappear. So I understand if we give up for now.

flomnes replied:
Right

flomnes committed:
Rename thermalPreproMax -> preproAvailabilityMax

flomnes committed:
Remove key_

payetvin committed:
code smells 2

payetvin committed:
Merge branch 'feature/links-gen-io' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into feature/links-gen-io

payetvin committed:
code smells 3, use std::array

payetvin committed:
Remove enums

payetvin committed:
Move arrays in func

payetvin committed:
Code smell 4

payetvin committed:
Fix reference array
